### PR TITLE
Allow BlockSyncer.listen to catch up with the node

### PR DESCRIPTION
**Pull request by rvl**
Relates to #52.

### Overview

It will now process blocks continuously, in chunks, until all the local blocks from the node are done. Then it will start polling.

### CODE REVIEW

File: test/unit/Cardano/NetworkLayerSpec.hs [hunk 1]
```diff
@@ -92,6 +94,9 @@ mkReader = do
         , \x -> modifyMVar_ ref $ return . (x :)
         )
 
+flatten :: [TickResult [Block]] -> [Block]
+flatten = concat . concat . map toList
```
Comment (paweljakubas) [inline test/unit/Cardano/NetworkLayerSpec.hs]: maybe : `concat . concatMap toList` ?

**Review by paweljakubas — APPROVED** [review]

File: test/unit/Cardano/NetworkLayerSpec.hs (hunk 1 above)
Comment (paweljakubas) [reply test/unit/Cardano/NetworkLayerSpec.hs]: oh, I see HLint complained about it

Comment (paweljakubas) [reply test/unit/Cardano/NetworkLayerSpec.hs]: oh I see HLint complained about it. I will fix this, squash and push

File: src/Cardano/NetworkLayer.hs [hunk 2]
```diff
@@ -39,24 +41,35 @@ data NetworkLayer m e0 e1 = NetworkLayer
...
+    case res of
+        GotChunk bs -> mapM_ action bs
+        Sleep -> liftIO $ threadDelay $ (fromIntegral . toMicroseconds) delay
     tick next action delay st'
 
+-- | The result type of the element fetch function provided to 'tick'.
+data TickResult a
+    = GotChunk !a -- ^ Have a result, and there may be more available.
+    | Sleep -- ^ There is no result available now, so wait.
+    deriving (Show, Eq, Foldable)
```
Comment (KtorZ) [inline src/Cardano/NetworkLayer.hs:71]: I wonder how the `Foldable` instance is derived here :thinking:

**Review by KtorZ — APPROVED** [review]

**Comment by KtorZ** [comment]
Please don't merge now. I am trying to resolve issues with building the http bridge first.

File: src/Cardano/NetworkLayer.hs (hunk 2 above)
Comment (paweljakubas) [reply src/Cardano/NetworkLayer.hs:71]:
```
{-# LANGUAGE DeriveFoldable #-}
```

**Comment by paweljakubas** [comment]
just rebased

**Pull request merged**